File: effs/words_read.js
// Words
// =====

// The whole file packed four bytes to a word, little-endian, in an array
// of 2^d words with at least three spare words of zeros, and its byte count.
function words_read(path) {
  const fs = require("fs");
  let buf;
  try {
    buf = fs.readFileSync(path);
  } catch (e) {
    return io_fail(Math.abs(e.errno ?? 2));
  }
  const words = (buf.length >> 2) + 3;
  let n = 1;
  while (n < words) {
    n *= 2;
  }
  const a = new Array(n).fill(0);
  for (let i = 0; i < buf.length; i += 4) {
    a[i >> 2] = ((buf[i] | 0) | ((buf[i + 1] | 0) << 8) | ((buf[i + 2] | 0) << 16) | ((buf[i + 3] | 0) << 24)) >>> 0;
  }
  return io_done(io_tup(a, buf.length));
}

File: effs/words_read.c
// Words
// =====
// the whole file packed four bytes to a word, little-endian, in a
// buf block of 2^d u32 lanes with at least three spare words of zeros,
// and its byte count
Term words_read_run(Env e, Term* f, IoWork* w) {
  uint64_t n = 0;
  char* path = io_cstr(e, f[0], &n);
  FILE* fp = io_nul(path, n) ? NULL : fopen(path, "rb");
  if (fp == NULL) {
    free(path);
    return io_fail(e, errno != 0 ? (u32)errno : ENOENT, NULL);
  }
  free(path);
  uint64_t cap = 1 << 20;
  uint64_t len = 0;
  uint8_t* buf = io_mem(malloc(cap));
  for (;;) {
    if (len + 65536 > cap) {
      cap *= 2;
      buf = io_mem(realloc(buf, cap));
    }
    size_t got = fread(buf + len, 1, cap - len, fp);
    if (got == 0) {
      break;
    }
    len += got;
  }
  fclose(fp);
  uint64_t words = (len >> 2) + 3;
  Nat d = 0;
  while ((1ull << d) < words) {
    d += 1;
  }
  Term arr = blk_new(e, 0, d, 0, 0, NULL);
  Loc loc = term_loc(arr);
  uint64_t full = len >> 2;
  for (uint64_t i = 0; i < full; i += 1) {
    uint32_t word;
    memcpy(&word, buf + i * 4, sizeof(word));
    blk_write(e.mem, 0, loc, (u32)i, word);
  }
  if ((len & 3) != 0) {
    uint32_t word = 0;
    memcpy(&word, buf + full * 4, len & 3);
    blk_write(e.mem, 0, loc, (u32)full, word);
  }
  free(buf);
  return io_done(e, io_tup(e, arr, (Term)len));
}

static void __attribute__((constructor)) words_read_use(void) {
  io_eff(CID_WORDS_READ, words_read_run, 0);
}
